File: solutions/c/collatz-conjecture/1/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start) {
    if (start < 1) return -1;

    int count = 0;
    for (; start > 1; count++) {
        start = start % 2 ? 3 * start + 1 : start / 2;
    }

    return count;
}

File: solutions/go/collatz-conjecture/1/collatz_conjecture.go
// Package collatzconjecture provides a function for the Collatz Conjecture.
package collatzconjecture

import "errors"

// CollatzConjecture returns the number of steps it takes to each 1 while
// applying the operations of the conjecture to a given positive integer
func CollatzConjecture(n int) (steps int, err error) {
	if n <= 0 {
		return 0, errors.New("not a positive integer")
	}

	for ; n > 1; steps++ {
		if n%2 == 0 {
			n /= 2
		} else {
			n = n*3 + 1
		}
	}

	return
}
